### ISSUE TITLE
Why not normalizing the gradient of perturbation

### ISSUE THREAD

**Issue by 4fee8fea** [devnkong/FLAG#9]
Thanks for your article and codes!

When I read `attack.py`, I notice there is no F-norm normalization as the paper said.

Indeed, `torch.sign` is used here.

Could you tell me the consideration behind, thank you very much.

```
def flag(model_forward, perturb_shape, y, args, optimizer, device, criterion) :
    model, forward = model_forward
    model.train()
    optimizer.zero_grad()

    perturb = torch.FloatTensor(*perturb_shape).uniform_(-args.step_size, args.step_size).to(device)
    perturb.requires_grad_()
    out = forward(perturb)
    loss = criterion(out, y)
    loss /= args.m

    for _ in range(args.m-1):
        loss.backward()
        perturb_data = perturb.detach() + args.step_size * torch.sign(perturb.grad.detach())
        perturb.data = perturb_data.data
        perturb.grad[:] = 0

        out = forward(perturb)
        loss = criterion(out, y)
        loss /= args.m

    loss.backward()
    optimizer.step()

    return loss, out
```

**Comment by devnkong** [comment]
Hey thanks for your interest in our paper!

Yep you can definitely try to use the F-norm normalization to do the job, here we just choose to use L-inf-norm-like normalization. I believe both can achieve the same level of performance but the sign() function is surely much faster.
